overlay.go

```diff
@@ -200,7 +208,19 @@ func visitFieldsInMap(ctx *visitorContext, val reflect.Value) (result bool) {
 			ctx.indexedElemPreVisitor(key)
 		}
 
-		result = visitField(ctx, v.Interface()) || result
+		visRes := visitField(ctx, v.Interface())
+		result = visRes || result
+
+		// If the map value is a non-scalar value and the visitor
+		// returned false (don't retain), consult the dropNonRequiredMapKeys
+		// hint to decide whether we need to delete the key from the map.
+		//
+		// This is required when splitting bundles into base/overlay
+		// bits as empty map values would be encoded as empty objects
+		// that the overlay merge code would mis-interpret as deletions.
+		if !visRes && isNonScalar(v) && ctx.dropNonRequiredMapKeys {
```

Comment:
I don't quite understand why this works. It feels like it should be !ctx.dropNonRequired...
because this is "I don't have a value, but the type isn't scalar". eg I have a map.
Which means we shouldn't set *anything* if we are wanting to drop non-required keys. Otherwise we will mark the keys for deletion.

Maybe I'm misinterpreting the definition of "dropNonRequiredMapKeys" ?

Reply:
The `visitField` implementation checks the type of `v` and invokes specialized methods for working with maps, slices or structs. Otherwise, it returns `false` which is interpreted as "I don't really know how to decide if you need scalar values so I just return false to be on the safe-side". This assumes that the parent container (map, slice, pointer to some object in the struct) is properly annotated so we can retain it if we need to (the offers map for instance)

An easy example to visualize this is to think of the struct processing code operating on a `BundleData` instance and currently processing the `ApplicationMap`. So eventually, the `visitFieldsInMap` gets called. We iterate the map and recursively visit each `*Application`. entry in the map. The first visitor returns true (it has an offer) and the second returns false. 

The predicate here should be interpreted as: `I don't need to retain the map entry for this key (visitor returned false) AND the type of the value is non-scalar AND I was asked to drop these kinds of keys from the map`. 

Essentially, the implementation assumes that the value which should probably not be retained points to an object. This means the struct processing code must have executed and we can safely trust the result given that only struct processing code checks for overlay-related annotations.

Hope my explanation helps to answer your question.

Reply:
It was just that I didn't really Set(..., reflect.Value{}) is actually Delete(). LGTM